tests/integration/test_charm.py

```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import pathlib
+
+import pytest
+import requests
+import tenacity
+import yaml
+from pytest_operator import plugin as pytest_plugin
+
+logger = logging.getLogger(__name__)
+
+LEGEND_HOST = "finos-legend"
+GITLAB_CLIENT_ID = "fake-client"
+GITLAB_CLIENT_SECRET = "fake-secret"
+
+METADATA = yaml.safe_load(pathlib.Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+
+MONGODB_APP_NAME = "mongodb-k8s"
+ENGINE_APP_NAME = "finos-legend-engine-k8s"
+SDLC_APP_NAME = "finos-legend-sdlc-k8s"
+STUDIO_APP_NAME = "finos-legend-studio-k8s"
+LEGEND_DB_APP_NAME = "finos-legend-db-k8s"
+GITLAB_INTEGRATOR_APP_NAME = "finos-legend-gitlab-integrator-k8s"
+NGINX_INGRESS_CHARM = "nginx-ingress-integrator"
+
+# TODO: Add STUDIO_APP_NAME to the list below once the following issue has been resolved:
+# https://github.com/finos/legend-studio/issues/1028
+LEGEND_APPS = [ENGINE_APP_NAME, SDLC_APP_NAME]
+LEGEND_APPS_CONFIG = [ENGINE_APP_NAME, SDLC_APP_NAME, STUDIO_APP_NAME]
+
+APP_LIST = [
+    MONGODB_APP_NAME,
+    LEGEND_DB_APP_NAME,
+    SDLC_APP_NAME,
+    ENGINE_APP_NAME,
+    STUDIO_APP_NAME,
+    GITLAB_INTEGRATOR_APP_NAME,
+]
+USEFUL_APPS = [
+    MONGODB_APP_NAME,
+    LEGEND_DB_APP_NAME,
+    ENGINE_APP_NAME,
+    STUDIO_APP_NAME,
+    GITLAB_INTEGRATOR_APP_NAME,
+]
+
+APP_PORTS = {
+    ENGINE_APP_NAME: 6060,
+    SDLC_APP_NAME: 7070,
+    STUDIO_APP_NAME: 8080,
+}
+APP_ROUTES = {
+    ENGINE_APP_NAME: "/engine",
+    SDLC_APP_NAME: "/api",
+    STUDIO_APP_NAME: "/studio",
+}
+
+
+@tenacity.retry(
+    retry=tenacity.retry_if_result(lambda x: x is False),
+    stop=tenacity.stop_after_attempt(10),
+    wait=tenacity.wait_exponential(multiplier=1, min=5, max=30),
+)
+def check_legend_connection(app_name, url, headers=None):
+    """Tests the connection to the given Legend URL.
+
+    When connecting to the FINOS Legend Applications, it will redirect to gitlab for
+    authentication, which will not be a 20X status code. We will receive 302 Found
+    instead.
+    Returns True if the response status code is 302, False in any other case.
+    """
+    logger.info("Trying to access %s...", app_name)
+    try:
+        response = requests.get(url, headers=headers, allow_redirects=False, timeout=10)
+        return response.status_code == 302
+    except Exception as ex:
+        logger.info(ex)
+    return False
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: pytest_plugin.OpsTest):
+    """Build the charm-under-test and deploy it together with related charms.
+
+    Assert on the unit status before any relations/configurations take place.
```

Comment:
no assertions are being made here, which makes this incorrect. I think we should leave only the first docstring line.

Reply:
Solved.